src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11041,12 +11101,25 @@ private LambdaExpressionSyntax ParseLambdaExpression(SyntaxToken asyncToken)
             else
             {
                 var name = this.ParseIdentifierToken();
-                var arrow = this.EatToken(SyntaxKind.EqualsGreaterThanToken);
-                arrow = CheckFeatureAvailability(arrow, MessageID.IDS_FeatureLambda);
-
+                // PROTOTYPE: TryEatToken
+                var exclamation = this.CurrentToken.Kind == SyntaxKind.ExclamationToken ? this.EatToken(SyntaxKind.ExclamationToken) : null;
+                SyntaxToken arrow;
+                if (this.CurrentToken.Kind == SyntaxKind.ExclamationEqualsToken && this.PeekToken(1).Kind == SyntaxKind.GreaterThanToken)
```

Comment:
i don't think this is necessarily right.  You're doing these checks after already eating the exclamation token above.  I would just enumerate the cases (like you did in higher parsing code) and just deal with each one.